// @visactor/vchart/cjs/chart/box-plot/box-plot-transformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.BoxPlotChartSpecTransformer = void 0;

const cartesian_1 = require("../cartesian"), util_1 = require("../util");

class BoxPlotChartSpecTransformer extends cartesian_1.CartesianChartSpecTransformer {
    _getDefaultSeriesSpec(spec) {
        var _a;
        const dataFields = [ spec.maxField, spec.medianField, spec.q1Field, spec.q3Field, spec.minField, spec.outliersField ], seriesSpec = Object.assign(Object.assign({}, super._getDefaultSeriesSpec(spec)), {
            boxPlot: spec.boxPlot,
            direction: null !== (_a = spec.direction) && void 0 !== _a ? _a : "vertical",
            minField: spec.minField,
            maxField: spec.maxField,
            q1Field: spec.q1Field,
            medianField: spec.medianField,
            q3Field: spec.q3Field,
            outliersField: spec.outliersField,
            outliersStyle: spec.outliersStyle
        });
        return seriesSpec["vertical" === seriesSpec.direction ? "yField" : "xField"] = dataFields, 
        seriesSpec;
    }
    transformSpec(spec) {
        super.transformSpec(spec), spec.axes || (spec.axes = [ {
            orient: "bottom"
        }, {
            orient: "left"
        } ]), (0, util_1.setDefaultCrosshairForCartesianChart)(spec);
    }
}

exports.BoxPlotChartSpecTransformer = BoxPlotChartSpecTransformer;
