// @visactor/vchart/esm/series/area/animation.js
import { registerLineAnimation, registerScaleInOutAnimation } from "../../animation/config";

import { Factory } from "../../core/factory";

const Appear_ClipIn = {
    type: "clipIn"
}, Appear_Grow = params => ({
    type: "horizontal" === params.direction ? "growPointsXIn" : "growPointsYIn",
    options: {
        orient: "horizontal" === params.direction ? "positive" : "negative"
    }
}), Appear_FadeIn = {
    type: "fadeIn"
};

export function areaPresetAnimation(params, preset) {
    if (!1 === preset) return {};
    switch (preset) {
      case "grow":
        return Appear_Grow(params);

      case "fadeIn":
        return Appear_FadeIn;

      default:
        return Appear_ClipIn;
    }
}

export const registerAreaAnimation = () => {
    Factory.registerAnimation("area", ((params, preset) => ({
        appear: areaPresetAnimation(params, preset),
        enter: {
            type: "fadeIn"
        },
        exit: {
            type: "fadeOut"
        },
        disappear: {
            type: "clipOut"
        }
    }))), registerLineAnimation(), registerScaleInOutAnimation();
};
