// @visactor/vchart/cjs/component/tooltip/handler/utils/position.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
}), exports.getVerticalPositionType = exports.getHorizontalPositionType = exports.positionType = exports.getActualTooltipPositionValue = void 0;

const vutils_1 = require("@visactor/vutils"), getActualTooltipPositionValue = (position, event) => {
    let result;
    if ((0, vutils_1.isValid)(position)) if ((0, vutils_1.isNumber)(position)) result = position; else if ((0, 
    vutils_1.isFunction)(position)) {
        const tooltipPosition = position(event);
        (0, vutils_1.isNumber)(tooltipPosition) && (result = tooltipPosition);
    }
    return result;
};

exports.getActualTooltipPositionValue = getActualTooltipPositionValue, exports.positionType = {
    left: [ "left", "middle" ],
    right: [ "right", "middle" ],
    inside: [ "middle", "middle" ],
    top: [ "middle", "top" ],
    lt: [ "left", "top" ],
    tl: [ "left", "top" ],
    rt: [ "right", "top" ],
    tr: [ "right", "top" ],
    bottom: [ "middle", "bottom" ],
    bl: [ "left", "bottom" ],
    lb: [ "left", "bottom" ],
    br: [ "right", "bottom" ],
    rb: [ "right", "bottom" ]
};

const getHorizontalPositionType = (position, defaultCase) => {
    var _a, _b;
    return null !== (_b = null === (_a = exports.positionType[position]) || void 0 === _a ? void 0 : _a[0]) && void 0 !== _b ? _b : defaultCase;
};

exports.getHorizontalPositionType = getHorizontalPositionType;

const getVerticalPositionType = (position, defaultCase) => {
    var _a, _b;
    return null !== (_b = null === (_a = exports.positionType[position]) || void 0 === _a ? void 0 : _a[1]) && void 0 !== _b ? _b : defaultCase;
};

exports.getVerticalPositionType = getVerticalPositionType;
