// @visactor/vchart/esm/component/tooltip/handler/utils/position.js
import { isFunction, isNumber, isValid } from "@visactor/vutils";

export const getActualTooltipPositionValue = (position, event) => {
    let result;
    if (isValid(position)) if (isNumber(position)) result = position; else if (isFunction(position)) {
        const tooltipPosition = position(event);
        isNumber(tooltipPosition) && (result = tooltipPosition);
    }
    return result;
};

export const positionType = {
    left: [ "left", "middle" ],
    right: [ "right", "middle" ],
    inside: [ "middle", "middle" ],
    top: [ "middle", "top" ],
    lt: [ "left", "top" ],
    tl: [ "left", "top" ],
    rt: [ "right", "top" ],
    tr: [ "right", "top" ],
    bottom: [ "middle", "bottom" ],
    bl: [ "left", "bottom" ],
    lb: [ "left", "bottom" ],
    br: [ "right", "bottom" ],
    rb: [ "right", "bottom" ]
};

export const getHorizontalPositionType = (position, defaultCase) => {
    var _a, _b;
    return null !== (_b = null === (_a = positionType[position]) || void 0 === _a ? void 0 : _a[0]) && void 0 !== _b ? _b : defaultCase;
};

export const getVerticalPositionType = (position, defaultCase) => {
    var _a, _b;
    return null !== (_b = null === (_a = positionType[position]) || void 0 === _a ? void 0 : _a[1]) && void 0 !== _b ? _b : defaultCase;
};
